### PR TITLE
Symlink support for packager

### PR DESCRIPTION
**Pull request by Kureev**
I saw we have quite a few user requests for symlink support...

**Test plan (required)**

1. Create a symlink in `node_modules` (for instance use `npm link`)
2. Run `npm start`
3. Profit!

**Code formatting**

Look around. Match the style of the rest of the codebase. See also the simple [style guide](https://github.com/facebook/react-native/blob/master/CONTRIBUTING.md#style-guide).

For more info, see the ["Pull Requests" section of our "Contributing" guidelines](https://github.com/facebook/react-native/blob/master/CONTRIBUTING.md#pull-requests).

### CODE REVIEW

**Comment by facebook-github-bot-8** [comment]
By analyzing the blame information on this pull request, we identified @philikon and @christopherdro to be potential reviewers.

**Comment by bestander** [comment]
Cc @matryoshcow, our new Packager man in RN team

**Comment by bestander** [comment]
FYI, big changes will be coming to packager, so all changes to current one need to be considered

**Comment by matryoshcow** [comment]
@Kureev, this is pretty neat! A couple of thoughts:
1) as it stands, this would be enabled by default, even though it's clearly targeted at specific class of users; it might be a good idea to make symlink support optional
2) to the extent possible, we're using async I/O; could your solution be modified in this manner?

bonus nitpick: one-letter variable names are the reason why we can't have nice things :)

**Comment by Kureev** [comment]
Hi @matryoshcow! 

About your notes:
1) Yeah, it's definitely not for everybody, but on the other hand the overhead you have is just about `fs.readdir` + `fs.lstat`. You can already specify --projectRoots and it'll work for you. My intention was to make it work "by default"
2) Sure!

**Comment by mkonicek** [comment]
@matryoshcow If I have no symlinks in node_modules, this feature is disabled. This way the feature is only enabled for people who want to use it.

It might happen that people have symlinks in node_modules without knowing about it though, which could be confusing. Printing info about the found symlinks on packager startup could help.

Making this optional with a packager argument would work too, but people would have to learn about the extra argument (Google -> StackOverflow etc., figuring out if info is up to date, extra work), and what if people launch the packager using Xcode? There are several different ways people launch the packager which makes it more work to make the feature super easy to use.

**Comment by mkonicek** [comment]
I'll request changes just for adding some debugging output so people know they're using this feature.

Agree with @matryoshcow's nit that some variable names could be made a little bit clearer, for example: 

    .filter(...)
    .map(symlinkPath => path.resolve(process.cwd(), fs.readlinkSync(symlinkPath)));

**Comment by mkonicek** [comment]
In short I agree with @Kureev that making this "just work" is a nice developer experience, compared to having to figure out how to enable the feature.

**Comment by bestander** [comment]
A concern that an extra walk through the node_modules folder slows down
packager start.
Any extra IO adds up especially in this case when it is doing "native" node
fs operations that are prone to throw ENOENT.

File: local-cli/server/server.js [hunk 1]
```diff
@@ -74,6 +75,9 @@ function _server(argv, config, resolve, reject) {
     ? argToArray(args.projectRoots)
     : config.getProjectRoots();
 
+  args.projectRoots = args.projectRoots
+    .concat(findSymlinksPaths(path.resolve(process.cwd(), 'node_modules')));
```
Comment (bestander) [inline local-cli/server/server.js]: This means 750 lstatSync operations during packager start.

**Comment by matryoshcow** [comment]
@mkonicek hmm... agreed - making this "just work" is a good idea, as long as it doesn't slow down the startup time. Since we're only talking about node_modules here the overhead may be negligible even for us, but let's confirm this before we commit to patching it in.

**Comment by matryoshcow** [comment]
@Kureev let's add this under a flag to begin with and perhaps log the symlink resolution time; we can enable it by default once we've observed it for a bit.

**Comment by mkonicek** [comment]
Right, `readdirSync` is not recursive so this should be fast but agree adding the feature behind a flag first is a good way to introduce it. Good idea with logging the time!

**Comment by respectTheCode** [comment]
When I first started using React Native I linked a few modules and spent hours trying to figure out why my hello world app wouldn't build. I know better now but my first guess was that it was an npm issue not a problem with React Native. Coming from a node background `npm link` is part of my normal workflow. I would challenge you to think about the first time developer experience. I understand that in bigger projects it could be beneficial to have symlinks disabled but I would argue that most projects are not big enough for it to have any impact. Logging the time it took would be a great way to indicate to the developer when they should consider disabling it.

**Comment by Kureev** [comment]
I've made a small benchmark.

_Code:_
```js
const path = require('path');
const fs = require('fs');

fs.readdirSync('./node_modules')
  .map(f => path.resolve('./node_modules', f))
  .filter(d => fs.lstatSync(d).isSymbolicLink())
  .map(s => path.resolve(process.cwd(), fs.readlinkSync(s)));
```

_Results:_
image

Amount of files in _node_modules:_
image

Honestly, I don't see a point of having it behind the flag, because you can achieve this by using `--projectRoots` flag already...

**Comment by bestander** [comment]
0.06 does not look bad at all.
Considering that you did your measurements I withdraw my objection :)

**Comment by Kureev** [comment]
I can make it all async and try again

**Comment by bestander** [comment]
Would be great, thanks.

**Comment by Kureev** [comment]
Async version:
```js
const path = require('path');
const fs = require('fs');

function findSymlinks(filePaths, cb) {
  let count = 0;
  let symlinks = [];

  filePaths.forEach(filePath =>
    fs.lstat(filePath, (err, stats) => {
      if (err) {
        cb(err, null);
      }

      if (stats.isSymbolicLink()) {
        symlinks.push(filePath);
      }

      count++;

      if (count === filePaths.length) {
        cb(null, symlinks);
      }
    })
  );
}

function resolveSymlinks(symlinks, cb) {
  let resolvedSymlinks = [];
  let count = 0;

  symlinks.forEach(symlink => {
    fs.readlink(symlink, (err, resolved) => {
      if (err) {
        cb(err, null);
      }

      resolvedSymlinks.push(resolved);

      if (count === symlinks.length) {
        cb(null, resolvedSymlinks);
      }
    });
  });
}

function resolveNPMSymlinks() {
  fs.readdir('./node_modules', (err, files) => {
    if (err) {
      throw err;
    }

    const filePaths = files.map(f => path.resolve('./node_modules', f));
    findSymlinks(filePaths, (err, symlinks) => {
      if (err) {
        throw err;
      }

      resolveSymlinks(symlinks, (err, resolved) => cb(err, resolved));
    });
  });
}

resolveNPMSymlinks();
```

But I have no idea how to measure it's performance, tbh 😞

**Comment by Kureev** [comment]
I'd keep it as it is now.

**Comment by matryoshcow** [comment]
If the cost of resolving all the symlinks is 60ms then it might not be the worst thing to go with the more straight-forward sync version and just have it enabled for everyone, no? @bestander, @mkonicek, what do you think?

**Comment by bestander** [comment]
Sounds good to me

**Comment by matryoshcow** [comment]
@Kureev, would you mind renaming those variables? After that, as far as I`m concerned this PR can be merged in its sync form.

**Comment by Kureev** [comment]
Awesome! Will update it ASAP ❤️

**Comment by facebook-github-bot-8** [comment]
@Kureev updated the pull request.

**Comment by Kureev** [comment]
Does it look better now?

**Comment by mkonicek** [comment]
Cool that it's so fast! Would you mind logging the symlinks so it's easy to see the feature is being used, and log the time taken to find symlinks just so we can check it in our CI?

**Comment by facebook-github-bot-8** [comment]
@Kureev updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@Kureev updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@Kureev updated the pull request.

**Comment by Kureev** [comment]
image

**Comment by facebook-github-bot-8** [comment]
@Kureev updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@Kureev updated the pull request.

File: local-cli/server/findSymlinksPaths.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const path = require('path');
+const fs = require('fs');
+
+module.exports = function findSymlinksPaths(lookupFolder) {
+  const timeStart = Date.now();
+  const folders = fs.readdirSync(lookupFolder);
+  const timeEnd = Date.now();
+  console.log(`Scanning ${folders.length} folders in ${lookupFolder} (${timeEnd - timeStart}ms)`);
```
Comment (matryoshcow) [inline local-cli/server/findSymlinksPaths.js]: Shiny! :D One final thought: you are only logging the directory read time; could you please include the symlink resolving in your timing as well? This is relevant too and may in fact be more costly than the simple read op.

File: local-cli/server/server.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ const chalk = require('chalk');
 const formatBanner = require('./formatBanner');
 const path = require('path');
 const runServer = require('./runServer');
+const findSymlinksPaths = require('./findSymlinksPaths');
```
Comment (matryoshcow) [inline local-cli/server/server.js]: Also, it appears you're requiring the symlink resolver but you're not actually using it.

**Comment by facebook-github-bot-8** [comment]
@Kureev updated the pull request.

File: local-cli/server/server.js (hunk 3 above)
Comment (Kureev) [inline local-cli/server/server.js]: Yes. But on the other hand there is no way to know if you're using symlinks or not.

Comment (matryoshcow) [inline local-cli/server/server.js]: What I mean is - you might actually want to call findSymlinkPaths to... you know... find the symlink paths and perhaps append them where they belong, no? :)

File: local-cli/server/findSymlinksPaths.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+module.exports = function findSymlinksPaths(lookupFolder) {
+  const timeStart = Date.now();
+  const folders = fs.readdirSync(lookupFolder);
+  const resolvedSymlinks = folders.map(folder => path.resolve(lookupFolder, folder))
+    .filter(folderPath => fs.lstatSync(folderPath).isSymbolicLink())
+    .map(symlink => path.resolve(process.cwd(), fs.readlinkSync(symlink)));
+  const timeEnd = Date.now();
+
+  console.log(`Scanning ${folders.length} folders in ${lookupFolder} (${timeEnd - timeStart}ms)`);
```
Comment (bestander) [inline local-cli/server/findSymlinksPaths.js]: Just a nit: `Scanning ${folders.length} folders for symlinks in ....`.
To make it clear

**Comment by facebook-github-bot-8** [comment]
@Kureev updated the pull request.

File: local-cli/server/server.js (hunk 3 above)
Comment (Kureev) [inline local-cli/server/server.js]: I'd say I want to return an array of resolved symlink paths from this function and apply them wherever it'll be needed.

**Comment by Kureev** [comment]
After merging master in, I accidentally removed `findSymlinkPaths` invokation. Fixed.

**Comment by facebook-github-bot-9** [comment]
@Kureev updated the pull request.

**Comment by matryoshcow** [comment]
Yeah that's what I meant :). Thanks!

**Comment by matryoshcow** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-5** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1059355014101804/int_phab) to review internal test results.

**Comment by matryoshcow** [comment]
Done and done! Thank you for taking the time to do this! ;)

**Comment by bestander** [comment]
Looks like the last commit broke e2e test

File: local-cli/server/server.js [hunk 5]
```diff
@@ -18,7 +18,10 @@ const findSymlinksPaths = require('./findSymlinksPaths');
  * Starts the React Native Packager Server.
  */
 function server(argv, config, args) {
-  args.projectRoots = args.projectRoots.concat(args.root);
+  args.projectRoots = args.projectRoots.concat(
```
Comment (bestander) [inline local-cli/server/server.js]: The tests started failing with this change

**Comment by bestander** [comment]
To reproduce locally: `node ./scripts/run-ci-e2e-tests.js --android --js`

File: local-cli/server/server.js [hunk 6]
```diff
@@ -12,12 +12,16 @@ const chalk = require('chalk');
...
+const findSymlinksPaths = require('./findSymlinksPaths');
 
 /**
  * Starts the React Native Packager Server.
  */
 function server(argv, config, args) {
-  args.projectRoots = args.projectRoots.concat(args.root);
+  args.projectRoots = args.projectRoots.concat(
+    args.root,
+    findSymlinkPaths(path.resolve(process.cwd(), 'node_modules'))
```
Comment (bestander) [inline local-cli/server/server.js]: Adding to project roots totally breaks `react-native run-android` command.

To reproduce:
```
react-native init TestApp
cd TestApp
react-native run-android
```
Error text:

~/work/react-native/packager ~

ENOENT: no such file or directory, scandir '/Users/bestander/work/react-native/packager/node_modules'

~
Process terminated. Press <enter> to close the window

**Comment by Kureev** [comment]
I'm very sorry for submitting a broken PR. I've tested it for iOS only and didn't notice that runIOS has a different packager logic than runAndroid. The issue was in a way I detect `node_modules`: for iOS it works if I say `path.resolve(process.cwd(), 'node_modules')`, because runIOS doesn't mutate a current dir (but seems that Android does when it spawning a packager script (see [this](https://github.com/facebook/react-native/blob/master/local-cli/runAndroid/runAndroid.js#L174))). So, to make it universal, I now use `path.resolve(__dirname, '..', '..', '..')`.

**Comment by Kureev** [comment]
I'll create a new PR

**Comment by ekryski** [comment]
@Kureev which version of Node are you using to symlink? I'm running RN v0.32-rc.0 with node v6.4.0 and my `fs.readlinkSync(symlink)` call looks like it has an extra `../` so it can't find my symlinked module.

I end up with `/Users/.nvm/versions/node/v6.4.0/lib/node_modules/rnui` when instead I should have `/Users/eric/.nvm/versions/node/v6.4.0/lib/node_modules/rnui`.

My thought would be that it is a NodeJS issue but just curious if you have seen the same??

**Comment by ekryski** [comment]
I think I've tracked it down. Looks like:

```js
return path.resolve(process.cwd(), fs.readlinkSync(symlink));
```

should actually be:

```js
return path.resolve(lookupFolder, fs.readlinkSync(symlink));
```

on [this line](https://github.com/facebook/react-native/blob/master/local-cli/server/findSymlinksPaths.js#L9)

**Comment by bestander** [comment]
@ekryski, would you create a PR and cc @Kureev and me for review?

**Comment by ekryski** [comment]
@bestander for sure. I'm leaning towards this:

```js
.map(symlink => path.resolve(lookupFolder, fs.readlinkSync(symlink)));
```

cool?

**Comment by bestander** [comment]
Whatever works :)

**Comment by aleclarson** [comment]
DAE need **recursive** symlink support? Many of my local packages use symlinks in their `node_modules`.

&nbsp;

The ideal solution might be to:

- resolve symlinks in the `node_modules` of any local package that is (directly or indirectly) imported by the project root(s)

- watch the `node_modules` of those local packages (in case symlinks are added/removed)

&nbsp;

Note: When I say "local package", I mean one that isn't installed via `npm install`.

**Comment by aleclarson** [comment]
Also, what if the symlink resolves into a symlink?

For example, I have a symlink in my React project that points to a symlink in `~/lib/node_modules` that points to the actual library.

**Comment by ryanlinnane** [comment]
What's the current status of this? I found that via using npm link, watchman broke because it referenced a symlink to a symlink. Moreover, I found that the path construction sometimes was a directory off, so changing the base folder to `lookupFolder` fixed up a few issues.

Code:
```javascript
module.exports = function findSymlinksPaths(lookupFolder) {
  const timeStart = Date.now();
  const folders = fs.readdirSync(lookupFolder);
  const resolvedSymlinks = folders.map(folder => path.resolve(lookupFolder, folder))
    .filter(folderPath => fs.lstatSync(folderPath).isSymbolicLink())
    .map(symlink => fs.realpathSync(path.resolve(lookupFolder, fs.readlinkSync(symlink))));
  const timeEnd = Date.now();
  console.log(`Scanning ${folders.length} folders for symlinks in ${lookupFolder} (${timeEnd - timeStart}ms)`);
  return resolvedSymlinks;
};
```

**Comment by aleclarson** [comment]
@ryanlinnane Check out #9792 and let me know if you still have issues.

**Comment by cpojer** [comment]
I agree we should figure out a path forward to support this both in RNP and in Jest. Give us some time to get our house in order please, as @bestander pointed out we are planning some pretty significant changes.

**Pull request closed without merging**